datafusion/core/src/optimizer/subquery_filter_to_join.rs

```diff
@@ -0,0 +1,334 @@
+                if regular_filters.is_empty() {
+                    Ok(new_input)
+                } else {
+                    Ok(utils::add_filter(new_input, &regular_filters))
+                }
+            }
+            _ => {
+                // Apply the optimization to all inputs of the plan
+                utils::optimize_children(self, plan, execution_props)
+            }
+        }
+    }
+
+    fn name(&self) -> &str {
+        "subquery_filter_to_join"
+    }
+}
+
+fn extract_subquery_filters(expression: &Expr, extracted: &mut Vec<Expr>) -> Result<()> {
+    utils::expr_sub_expressions(expression)?
+        .into_iter()
+        .try_for_each(|se| match se {
+            Expr::InSubquery { .. } => {
+                extracted.push(se);
+                Ok(())
+            }
+            _ => extract_subquery_filters(&se, extracted),
+        })
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::logical_plan::{
+        and, binary_expr, col, in_subquery, lit, not_in_subquery, LogicalPlanBuilder,
+        Operator,
+    };
+    use crate::test::*;
+
+    fn assert_optimized_plan_eq(plan: &LogicalPlan, expected: &str) {
+        let rule = SubqueryFilterToJoin::new();
+        let optimized_plan = rule
+            .optimize(plan, &ExecutionProps::new())
+            .expect("failed to optimize plan");
+        let formatted_plan = format!("{}", optimized_plan.display_indent_schema());
+        assert_eq!(formatted_plan, expected);
+    }
+
+    fn test_subquery() -> Result<Arc<LogicalPlan>> {
+        let table_scan = test_table_scan()?;
+        Ok(Arc::new(
+            LogicalPlanBuilder::from(table_scan)
+                .project(vec![col("c")])?
+                .build()?,
+        ))
+    }
+
+    /// Test for single IN subquery filter
+    #[test]
+    fn in_subquery_simple() -> Result<()> {
+        let table_scan = test_table_scan()?;
+        let plan = LogicalPlanBuilder::from(table_scan)
+            .filter(in_subquery(col("c"), test_subquery()?))?
+            .project(vec![col("test.b")])?
+            .build()?;
+
+        let expected = "Projection: #test.b [b:UInt32]\
+        \n  Semi Join: #test.c = #test.c [a:UInt32, b:UInt32, c:UInt32]\
+        \n    TableScan: test projection=None [a:UInt32, b:UInt32, c:UInt32]\
+        \n    Projection: #test.c [c:UInt32]\
+        \n      TableScan: test projection=None [a:UInt32, b:UInt32, c:UInt32]";
+
+        assert_optimized_plan_eq(&plan, expected);
+        Ok(())
+    }
+
+    /// Test for single NOT IN subquery filter
+    #[test]
+    fn not_in_subquery_simple() -> Result<()> {
+        let table_scan = test_table_scan()?;
+        let plan = LogicalPlanBuilder::from(table_scan)
+            .filter(not_in_subquery(col("c"), test_subquery()?))?
+            .project(vec![col("test.b")])?
+            .build()?;
+
+        let expected = "Projection: #test.b [b:UInt32]\
+        \n  Anti Join: #test.c = #test.c [a:UInt32, b:UInt32, c:UInt32]\
```

Comment:
I can't tell from this plan if the predicate is correct (because `#test` is used as the relation name in both the inner and outer query. 

It might make these tests more readable if the relation name in the subquery was something different (like `test_sq`) so that this join predicate appears as `#test.c = #sq_ test.c`

Reply:
Done - now table names in tests are different, so it should by much easier to read